### PR TITLE
primaryattack tweaks

### PR DESCRIPTION
**Pull request by Jackarunda**
- put trybustdoor into its own func so that it can be easily overridden with different door breaching behavior
- changed MyEmitSound to have another argument that lets you use the chad sound.Play instead of the virgin EmitSound
- added a buff that can override the shell effect
- added a way to tweak shell eject positions
- fixed a rocket velocity bug
- generally added the code necessary for FireRocket() to actually be useful instead of a garbage func that was never tested

### CODE REVIEW

File: lua/weapons/arccw_base/sh_firing.lua [hunk 1]
```diff
@@ -300,6 +300,7 @@ function SWEP:PrimaryAttack()
     end
 
     self:DoShootSound()
+	if self.DoExtraFiringEvents then self:DoExtraFiringEvents() end
```
Comment (HaodongMo) [inline lua/weapons/arccw_base/sh_firing.lua:303]: Do this sort of thing using Hooks. Use self:GetBuff_Hook("Hook_ExtraFiringEvents") for example.

File: lua/weapons/arccw_base/sh_util.lua [hunk 2]
```diff
@@ -2,12 +2,16 @@ function SWEP:TableRandom(table)
...
-function SWEP:MyEmitSound(fsound, level, pitch, vol, chan)
+function SWEP:MyEmitSound(fsound, level, pitch, vol, chan, useWorld)
     fsound = self:GetBuff_Hook("Hook_TranslateSound", fsound) or fsound
 
     if istable(fsound) then fsound = self:TableRandom(fsound) end
 
     if fsound != "" then
-        self:EmitSound(fsound, level, pitch, vol, chan or CHAN_AUTO)
+		if useWorld then
+			sound.Play(fsound, self:GetPos() + VectorRand(), level, pitch, vol)
```
Comment (HaodongMo) [inline lua/weapons/arccw_base/sh_util.lua:12]: Why the VectorRand? Also, I'm not sure what self:GetPos() would result in, but you should probably do owner:EyePos()

File: lua/weapons/arccw_base/shared.lua [hunk 3]
```diff
@@ -183,13 +183,15 @@ SWEP.EnterBipodSound = "weapons/arccw/m249/m249_coverdown.wav"
...
 SWEP.ExitUBGLSound = "weapons/arccw/ubgl_exit.wav"
+SWEP.ShootSoundWorldCount = 0 -- every shot, play the fire sound this many more times in the world (adds volume), unless suppressed
 
 SWEP.NoFlash = nil -- disable light flash
 SWEP.MuzzleEffect = nil
 SWEP.FastMuzzleEffect = nil
 SWEP.GMMuzzleEffect = false -- Use Gmod muzzle effects rather than particle effects
 SWEP.ImpactEffect = nil
 SWEP.ImpactDecal = nil
+SWEP.DoExtraFiringEvents = nil -- override this function to add custom firing events
```
Comment (HaodongMo) [inline lua/weapons/arccw_base/shared.lua:194]: Again, should be a hook to allow attachments to do this stuff

File: lua/weapons/arccw_base/shared.lua [hunk 4]
```diff
@@ -183,13 +183,15 @@ SWEP.EnterBipodSound = "weapons/arccw/m249/m249_coverdown.wav"
 SWEP.ExitBipodSound = "weapons/arccw/m249/m249_coverup.wav"
 SWEP.SelectUBGLSound =  "weapons/arccw/ubgl_select.wav"
 SWEP.ExitUBGLSound = "weapons/arccw/ubgl_exit.wav"
+SWEP.ShootSoundWorldCount = 0 -- every shot, play the fire sound this many more times in the world (adds volume), unless suppressed
```
Comment (HaodongMo) [inline lua/weapons/arccw_base/shared.lua:186]: strange feature

**Review by HaodongMo — CHANGES_REQUESTED** [review]

File: lua/weapons/arccw_base/sh_firing.lua (hunk 1 above)
Comment (Jackarunda) [reply lua/weapons/arccw_base/sh_firing.lua:303]: if i do this using hooks, will i still be able to specify this behavior for certain weapons innately? i dont want to have to put like a ghost attachment on, say, the rocket launchers in order for them to have backblast

File: lua/weapons/arccw_base/sh_util.lua (hunk 2 above)
Comment (Jackarunda) [reply lua/weapons/arccw_base/sh_util.lua:12]: sure

File: lua/weapons/arccw_base/shared.lua (hunk 4 above)
Comment (Jackarunda) [reply lua/weapons/arccw_base/shared.lua:186]: perhaps, but it works, is disabled by default, and i need it for the jmod guns

Comment (Fesiug) [reply lua/weapons/arccw_base/shared.lua:186]: can't you do an hook_addshootsound and copy the values from the hook?

File: lua/weapons/arccw_base/sh_firing.lua (hunk 1 above)
Comment (Fesiug) [reply lua/weapons/arccw_base/sh_firing.lua:303]: nope, weapons can use the same hooks attachments use!
image

**Review by HaodongMo — APPROVED** [review]

**Pull request merged**